src/components/interactive-map.js

```diff
@@ -125,153 +124,178 @@ const defaultProps = Object.assign(
+
+    if (isEntering) {
+      onEvent.call(this, 'onMouseEnter', event);
+    }
+    if (isExiting) {
+      onEvent.call(this, 'onMouseLeave', event);
+    }
+    if (isEntering || isExiting) {
+      this.setState({isHovering});
     }
-    return features;
   }
+}
 
-  _onInteractionStateChange = interactionState => {
-    const {isDragging = false} = interactionState;
-    if (isDragging !== this.state.isDragging) {
-      this._updateInteractiveContext({isDragging});
-      this.setState({isDragging});
-    }
+function onPointerClick(event) {
+  const {onClick, onNativeClick, onDblClick, doubleClickZoom} = this.props;
+  let callbacks = [];
+  const isDoubleClickEnabled = onDblClick || doubleClickZoom;
+
+  // `click` is only fired on single click. `anyclick` is fired twice if double clicking.
+  // `click` has a delay period after pointer up that prevents it from firing when
+  // double clicking. `anyclick` is always fired immediately after pointer up.
+  // If double click is turned off by the user, we want to immediately fire the
+  // onClick event. Otherwise, we wait to make sure it's a single click.
+  switch (event.type) {
+    case 'anyclick':
+      callbacks.push(onNativeClick);
+      if (!isDoubleClickEnabled) {
+        callbacks.push(onClick);
+      }
+      break;
 
-    const {onInteractionStateChange} = this.props;
-    if (onInteractionStateChange) {
-      onInteractionStateChange(interactionState);
-    }
-  };
+    case 'click':
+      if (isDoubleClickEnabled) {
+        callbacks.push(onClick);
+      }
+      break;
 
-  _updateInteractiveContext(updatedContext) {
-    this._interactiveContext = Object.assign({}, this._interactiveContext, updatedContext);
+    default:
   }
 
-  _onResize = ({width, height}) => {
-    this._width = width;
-    this._height = height;
-    this._setControllerProps(this.props);
-    this.props.onResize({width, height});
+  callbacks = callbacks.filter(Boolean);
+
+  if (callbacks.length) {
+    event = normalizeEvent.call(this, event);
+    // backward compatibility: v3 `onClick` interface
+    event.features = getFeatures.call(this, event.point);
+    callbacks.forEach(cb => cb(event));
+  }
+}
+/* End of event handers */
+
+const InteractiveMap = forwardRef((props, ref) => {
+  const [controller] = useState(() => props.controller || new MapController());
+  const [eventManager] = useState(() => new EventManager(null, {touchAction: props.touchAction}));
+  const [context, setContext] = useState({
+    eventManager,
+    isDragging: false
+  });
+  const eventCanvasRef = useRef(null);
+  const staticMapRef = ref || useRef(null);
+
+  // Event handlers are registered once but need access to the latest props
+  // This is an anti-pattern, though it maintains a persistent reference to the latest props/state of this component
+  const _thisRef = useRef({
+    width: 0,
+    height: 0,
+    state: {
+      isHovering: false,
+      isDragging: false
+    }
+  });
```

Comment:
thank you for sharing this, i didn't know about this pattern. Based on this [reading](https://www.emgoto.com/storing-values-with-useref/) it looks like storing values in useRef doesn't trigger a re-render

Reply:
That is correct. The problem I'm trying to solve here is that event listeners are registered once but need access to the latest state in the handlers. We want to avoid rerender when this object is mutated. Its only purpose is to provide pointers to the current props/state when events are fired.